### PR TITLE
Refactor Teleport

### PR DESCRIPTION
**Pull request by fmatthew5876**
~~Depends on #1687~~

Fix: #1624
Fix: #1693 
Fix: #1316
Fix: #1567

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This passes all the test cases and is ready for review.

Now some trivia. If you `MoveToVariableLocation` to (-1,8) in RPG_RT on horizontal looping map, your position will look correct (as if you're at x=10) but the `position_x` chunk has the value of -1. :exclamation: Step left or right and x will fix itself to the correct looped value.

Not sure what kind of bugs that might cause with collisions in RPG_Rt.

This RPG_RT bug is fixed in Player.

**Comment by carstene1ns** [comment]
Better diff: https://github.com/fmatthew5876/Player/compare/event_makeway...fmatthew5876:event_teleport

**Comment by Ghabry** [comment]
This causes multiple issues:

https://easyrpg.org/play/pr1694/?game=yume_nikki
After the instructions the screen stays black.

https://easyrpg.org/play/pr1694/
Talk to the dude next to the bridge, the message box closes after the teleport.

**Comment by fmatthew5876** [comment]
I fixed the test game issue, but the Yume Nikki thing is still outstanding.

It only happens on web, and not on my local build in linux. The first "Refactor Teleport" commit moved things around and touched some code with comments about emscripten, so this is probably a web only regression.

@Ghabry What exactly is the issue here with async loading and transitions that breaks emscripten?

**Comment by fmatthew5876** [comment]
I've fixed the hang in Yume Niiki. But now there is another issue. 

The intro is supposed to fade in smoothly with "Working... Please Wait..."

What happens instead is a fade in screen transition, and then the black picture pops immediately after.

The intro has this autostart code:
```
Transfer Player Map3
ShowPicture 4 (all black)
ShowPicture 5 (Working... Please Wait...)
Show Screen
MovePicture4 Transparent 0% -> 100% 1.0s
MovePicture5 Transparent 0% -> 100% 1.0s Wait
```

In emscripten, the pictures don't finish async loading until after the transition for show screen starts. When the transition starts it calls `Transition::Init()`, and this is what snaps the transition end bitmap without the pictures since they haven't loaded yet.

I'm not sure this is actually a "bug" but maybe a side effect of calling the update routines and interpreters at the right time after teleport.

To fix this, we should somehow need to block on any graphical assets aysnc loading before starting a transition.

My log shows this:
```
Debug: PRE UPDATE pr1694:52:11
Debug: MAP UPD 3 1 pr1694:52:11
Debug: FORCE ERASE SKIP pr1694:52:11
Debug: MAP UPD 3 0 pr1694:52:11
Debug: COMMAND SHOW PIC pr1694:52:11
Debug: BIND FUNC OBJ pr1694:52:11
Debug: COMMAND SHOW PIC pr1694:52:11
Debug: BIND FUNC OBJ pr1694:52:11
Debug: SHOW SCREEN pr1694:52:11
Debug: CALL LISTENERS pr1694:52:11
Debug: USER TRANSITION START pr1694:52:11
Debug: MAP UPD 3 0 pr1694:52:11
Debug: CALL LISTENERS pr1694:52:11
Debug: PIC SPRITE READY pr1694:52:11
Debug: CALL LISTENERS pr1694:52:11
Debug: PIC SPRITE READY
```

Regardless, the fix for async was complex and nasty. I now need to retest all the teleport cases and pre-update scene calling again to make sure I didn't break it.

**Comment by Ghabry** [comment]
Yeah you need one delay-frame, otherwise emscripten will load black because it has no time to do "blocking asset loading". THis was all discussed in the comments you deleted:

```
// Fade Out not handled here but in StartTeleportPlayer because otherwise 
// emscripten hangs before fading out when doing async loading... 
 
// Async loading note: 
// Fade In must be done before finish teleport, otherwise chipset is not 
// loaded and renders black while fading -> ugly 
 
// The delay is only needed for one frame to execute pending transitions, 
// the interpreters continue on the old map afterwards 
```

You can fake async loading on PC by changing to line in asynchandler.cpp:

Remove the comment after ``// remove comment for fake download testing`` and comment the line after ``// add comment for fake download testing``. Then you get a fake delay which simulates async on a bad connection

**Comment by fmatthew5876** [comment]
> Yeah you need one delay-frame,

Only one delay frame is enough? What if the asset takes more than 1 frame time to download? Or will emscripten block calling the update loop when there are pending loads?

**Comment by Ghabry** [comment]
That's what the "IsImportantFilePending" check does.
When any Request has the "SetImportant(true)" set the Updates won't be called.
(that's how Database and Map downloads work, they set this flag).

You could e.g. set it on any asset that is requested while the screen is faded out, then the fade-in won't happen until the Transitions are done.

**Comment by fmatthew5876** [comment]
Async issues with Yume Nikki are now fixed.

I have to redo the teleport test cases now. But this can now be properly reviewed at the same time.

I think we have a path forward with transitions from this, which is great news. That'll be done proper in a future PR.

**Comment by fmatthew5876** [comment]
I've redone all the test cases and fixed some issues.

This is ready now.

**Comment by Ghabry** [comment]
woah the yume nikki fadein is for the first time looking correct. great.

File: src/teleport_target.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		int GetDirection() const;
+
+		/** @return whether this is active */
+		bool IsActive() const;
+	private:
+		int _map_id = 0;
+		int _x = 0;
+		int _y = 0;
+		int16_t _d = -1;
+		bool _active = false;
```
Comment (Ghabry) [inline src/teleport_target.h]: minor nitpick: We don't use vars prefixed with _

**Comment by Ghabry** [comment]
Tested #1416: Works.

Log.[in] seems to be broken in the web player in general (in all branches), not tested there.

**Comment by fmatthew5876** [comment]
I fixed the style comments

**Comment by carstene1ns** [comment]
ok

**Pull request merged**